src/v/kafka/security/scram_authenticator.cc

```diff
@@ -0,0 +1,119 @@
+ */
+#include "kafka/security/scram_authenticator.h"
+
+#include "vlog.h"
+
+namespace kafka {
+
+template<typename T>
+bytes scram_authenticator<T>::handle_client_first(bytes_view auth_bytes) {
+    // request from client
+    _client_first = std::make_unique<client_first_message>(auth_bytes);
+    vlog(klog.debug, "Received client first message {}", *_client_first);
+
+    // lookup credentials for this user
+    _authid = _client_first->username_normalized();
+    auto credential = _credentials.get<scram_credential>(_authid);
+    if (!credential) {
+        throw sasl_error(
+          fmt::format("Invalid auth user credentials: {}", _authid));
+    }
+    _credential = std::make_unique<scram_credential>(std::move(*credential));
+
+    if (
+      !_client_first->authzid().empty()
+      && _client_first->authzid() != _authid) {
+        throw sasl_error(fmt::format("Invalid auth id and username pair"));
+    }
+
+    if (_credential->iterations() < scram::min_iterations) {
+        throw sasl_error(fmt::format(
+          "Requested iterations {} less than minimum {}",
+          _credential->iterations(),
+          scram::min_iterations));
+    }
+
+    // build server reply
+    _server_first = std::make_unique<server_first_message>(
+      _client_first->nonce(),
+      random_generators::gen_alphanum_string(nonce_size),
+      _credential->salt(),
+      _credential->iterations());
+
+    _state = state::client_final_message;
+
+    auto reply = _server_first->sasl_message();
+    return bytes(reply.cbegin(), reply.cend());
+}
+
+template<typename T>
+bytes scram_authenticator<T>::handle_client_final(bytes_view auth_bytes) {
+    client_final_message client_final(auth_bytes);
+    vlog(klog.debug, "Received client final message {}", client_final);
+
+    auto client_signature = scram::client_signature(
+      _credential->stored_key(), *_client_first, *_server_first, client_final);
+
+    auto computed_stored_key = scram::computed_stored_key(
+      client_signature,
+      bytes(client_final.proof().begin(), client_final.proof().end()));
+
+    if (computed_stored_key != _credential->stored_key()) {
+        throw sasl_error("Invalid auth credentials from client");
+    }
+
+    const auto username = _client_first->username();
+    vlog(klog.debug, "Authentication key match for user {}", username);
+
+    auto server_signature = scram::server_signature(
+      _credential->server_key(), *_client_first, *_server_first, client_final);
+
+    server_final_message server_final(std::nullopt, server_signature);
+    auto reply = server_final.sasl_message();
+
+    clear_credentials();
+    _state = state::complete;
+
+    return bytes(reply.cbegin(), reply.cend());
+}
+
+template<typename T>
+void scram_authenticator<T>::clear_credentials() {
+    _credential.reset();
+    _client_first.reset();
+    _server_first.reset();
+}
+
+template<typename T>
+bytes scram_authenticator<T>::authenticate(bytes_view auth_bytes) {
+    try {
```

Comment:
should this be `result<bytes>` and int error codes?

Reply:
what do you have in mind? the counter point is that effectively all errors have the same effect: auth is failed. the only thing that matters is a difference in error messages.

Reply:
exceptions unhandled crash the program. `result<>` tells the user that an error is expected. 

i think from an API design perspective, a `result<T>` communicates the intent better to me.

Reply:
can result<> communicate a string error? because there aren't multiple numeric error codes; only one: failed.

Reply:
i don't thinkso, for that you'd need a variant of sstring + bytes which actually would be a 0 cost variant since both bytes and sstring use the exact same technique of short string optimization. 

however, the alternative is to enumerate the failures, which doesn't seem like that many (10?) - this is what RPC did for example. 

turns out that there weren't that many erorr paths there when we moved to result<>

i guess i was trying to say that all parsing code is really prone to bad parsing errors/malformed bytes.

Reply:
yeh worked out nicely